pkg/reconciler/v1alpha1/route/resources/cluster_ingress.go

```diff
@@ -0,0 +1,179 @@
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"fmt"
+	"sort"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/knative/pkg/kmeta"
+	"github.com/knative/serving/pkg/activator"
+	"github.com/knative/serving/pkg/apis/networking/v1alpha1"
+	"github.com/knative/serving/pkg/apis/serving"
+	servingv1alpha1 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	"github.com/knative/serving/pkg/reconciler"
+	revisionresources "github.com/knative/serving/pkg/reconciler/v1alpha1/revision/resources"
+	"github.com/knative/serving/pkg/reconciler/v1alpha1/route/resources/names"
+	"github.com/knative/serving/pkg/reconciler/v1alpha1/route/traffic"
+	"github.com/knative/serving/pkg/system"
+)
+
+// MakeClusterIngress creates ClusterIngress to set up routing rules. Such ClusterIngress specifies
+// which Hosts that it applies to, as well as the routing rules.
+func MakeClusterIngress(r *servingv1alpha1.Route, tc *traffic.TrafficConfig) *v1alpha1.ClusterIngress {
+	return &v1alpha1.ClusterIngress{
+		ObjectMeta: metav1.ObjectMeta{
+			// As ClusterIngress resource is cluster-scoped,
+			// here we use GenerateName to avoid conflict.
+			GenerateName: fmt.Sprintf("%s-", names.ClusterIngress(r)),
+			Labels: map[string]string{
+				serving.RouteLabelKey:          r.Name,
+				serving.RouteNamespaceLabelKey: r.Namespace,
+			},
+			OwnerReferences: []metav1.OwnerReference{*kmeta.NewControllerRef(r)},
+		},
+		Spec: makeClusterIngressSpec(r, tc.Targets),
+	}
+}
+
+func makeClusterIngressSpec(r *servingv1alpha1.Route, targets map[string][]traffic.RevisionTarget) v1alpha1.IngressSpec {
+	// Domain should have been specified in route status
+	// before calling this func.
+	domain := r.Status.Domain
+	names := []string{}
+	for name := range targets {
+		names = append(names, name)
+	}
+	// Sort the names to give things a deterministic ordering.
+	sort.Strings(names)
+	// The routes are matching rule based on domain name to traffic split targets.
+	rules := []v1alpha1.ClusterIngressRule{}
+	for _, name := range names {
+		rules = append(rules, *makeClusterIngressRule(getRouteDomains(name, r, domain), r.Namespace, targets[name]))
+	}
+	return v1alpha1.IngressSpec{
+		Rules: rules,
+	}
+}
+
+func getRouteDomains(targetName string, r *servingv1alpha1.Route, domain string) []string {
+	var domains []string
+	if targetName == "" {
+		// Nameless traffic targets correspond to many domains: the
+		// Route.Status.Domain, and also various names of the Route's
+		// headless Service.
+		domains = []string{domain,
```

Comment:
ha, I made it this way mainly because I don't expect to see dedup everywhere.
It's a minor issue on code style preference anyway.
Since you have strong persistence on code conciseness, Let's make it the way you suggested :)

Reply:
Ah, I see what you mean.
The domains do not indeed need `dedup` for named target, so it's fine.